### PR TITLE
Segmentation 3D example ported to substra

### PR DESCRIPTION
**Pull request by jmorel**
Check the README.

# Randomization issue with openers

`substratools.Opener` requests that user implement both `get_X` and `get_y`. It gets really tricky when using a randomized batch iterator like DataLoader because we cannot instantiate 2 DataLoader (one in each method) on the same data samples because each loader would be randomized differently and `X_i` would not match `y_i`.

Also we cannot use a randomized loader for testtuple, otherwise the order of items in y_pred (generated from one instance of the loader) would not match the one in y_true (generated from another instance of the loader).

A good solution would be to be able to set the order of data samples in the testtuple but this is currently not supported by substra. This way the randomization could be done outside of the opener.

Note: the randomization is especially important when trying to run multiple epochs (with randomization happing in-between epochs). In substra however each epoch should be a separate traintuple. So it would be perfectly possible to randomize differently the order of data samples at each traintuple.

In the meantime, here are a few solutions:

1. Have a separate opener for training and one for testing. This is necessary if we want to keep the randomization.

2. Only load the data once during training, either by:
  a. having get_y return None and get_X return the loader (implemented here)
  b. having get_y and get_X return iterators that iterate over separate parts of the loader (can be tricky)

# About run-local

Attempting to run the example using `substra run-local` raise a memory error. By patching the code of the runner in the substra repo as follows I was able to run it without trouble. The patch is inspired by code found in substra-backend.

```diff
diff --git a/substra/runner.py b/substra/runner.py
index d995950..42b9d9b 100644
--- a/substra/runner.py
+++ b/substra/runner.py
@@ -21,6 +21,7 @@ import tempfile
 import time
 import hashlib
 import zipfile
+from subprocess import check_output

 import docker

@@ -84,6 +85,35 @@ def _docker_build(docker_client, dockerfile_path, name, rm=False):
 def _docker_run(docker_client, name, command, volumes, remove=True):
     print(f'Running docker {name}', end=' ', flush=True)
     start = time.time()
+
+    # memory_limit_mb = f'{resources_manager.memory_limit_mb()}M'
+    # cpu_set, gpu_set = resources_manager.get_cpu_gpu_sets()    # blocking call
+
+    try:
+        memory_limit_mb = int(os.sysconf('SC_PAGE_SIZE') * os.sysconf('SC_PHYS_PAGES') / (1024. ** 2))
+    except ValueError:
+        memory_limit_mb = int(check_output(['sysctl', '-n', 'hw.memsize']).strip())
+
+    task_args = {
+        'image': name,
+        # 'name': container_name,
+        # 'cpuset_cpus': cpu_set,
+        'mem_limit': f"{memory_limit_mb}M",
+        'command': command,
+        'volumes': volumes,
+        'shm_size': '8G',
+        # 'labels': [DOCKER_LABEL],
+        'detach': False,
+        'stdout': True,
+        'stderr': True,
+        'auto_remove': False,
+        'remove': False,
+        'network_disabled': True,
+        'network_mode': 'none',
+        'privileged': False,
+        'cap_drop': ['ALL'],
+        # 'environment': environment
+    }
     try:
         # Setting userns_mode to "host" effectively turns off user namespaces
         # (see https://github.com/moby/moby/issues/25492#issuecomment-239173095).
@@ -91,9 +121,10 @@ def _docker_run(docker_client, name, command, volumes, remove=True):
         # filesystem from the container.
         # It is safe to do because we also use the user=USER option: the `UID`
         # in the container is set to the `UID` of the current process.
-        docker_client.containers.run(name, command=command,
-                                     volumes=volumes, remove=remove, user=USER,
-                                     userns_mode="host")
+        # docker_client.containers.run(name, command=command,
+        #                              volumes=volumes, remove=remove, user=USER,
+        #                              userns_mode="host")
+        docker_client.containers.run(**task_args)
     except docker.errors.ContainerError as e:
         # try to pretty print traceback
         try:
```

### CODE REVIEW

File: examples/segmentation_3d/assets/algo/algo.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+            in_channels=1,
+            out_channels=1,
+            channels=(16, 32, 64, 128, 256),
+            strides=(2, 2, 2, 2),
+            num_res_units=2,
+        ).to(self.device)
+        optimizer = torch.optim.Adam(model.parameters(), 1e-3)
+        return (model, optimizer)
+
+    def train(self, X, y, models, rank):
```
Comment (samlesu) [inline examples/segmentation_3d/assets/algo/algo.py]: It seems here, that the labels are encoded inside each x item.

This is a bit of a hack: it means that you have access in the predict function to the associated labels.

File: examples/segmentation_3d/assets/algo/algo.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    def predict(self, X, model):
+        y_pred = []
+        model, _ = model  # drop the optimizer
+        model.eval()
+        with torch.no_grad():
+            for inputs, metadata in X:
+                roi_size = (96, 96, 96)
+                sw_batch_size = 4
+                outputs = sliding_window_inference(inputs, roi_size, sw_batch_size, model)
+                y_pred.append((outputs, metadata))
```
Comment (samlesu) [inline examples/segmentation_3d/assets/algo/algo.py]: Could you explain why you are returning both the predicted outputs and the input labels?

(It seems not used inside the score function)

File: examples/segmentation_3d/assets/train_dataset/opener.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+from glob import glob
+
+from monai.data import NiftiDataset
+from monai.transforms import Compose, AddChannel, ScaleIntensity, ToTensor, RandSpatialCrop, RandRotate90
+import torch
+from torch.utils.data import DataLoader
+
+import substratools as tools
+
+class MonaiTrainOpener(tools.Opener):
```
Comment (samlesu) [inline examples/segmentation_3d/assets/train_dataset/opener.py]: Could you please explain why we need two opener (for testing and training) ?

**Review by samlesu — COMMENTED** [review]
Thanks @jmorel; looks promising!

File: examples/segmentation_3d/assets/algo/algo.py (hunk 1 above)
Comment (jmorel) [reply examples/segmentation_3d/assets/algo/algo.py]: Monai provides a DatasetLoader that randomizes input data and groups it in batches. This loader is an iterator over the batches. Because of the randomization, we cannot load X and y separately, otherwise matching values in X and y could be at different indexes and in different batches.

But the train and eval step use different loading strategy: randomized batches of 5 items for train, and non-randomized batches of 1 item for eval. My solution was therefore to have 2 openers.

For train, the loader would be returned by get_X (get_y returning None) and the train method would have to iterate over the X args to get the actual X and y values.

For test however, I made 2 generators that iterate of the loader, one for get_X and one for get_y. Since the loader is non-randomized, each instance of the loader returns the data in the same order. By adding generators on top of the loader that drop in one case the X value and in the other one the y value, we're able to load X and y separately.

To make sure that the train opener cannot be used for predictions, its save_prediction method raise an exception. This way any testtuple referencing this opener will fail.

You could however use the test opener for training, but it would quite slow as there is no batching and data is loaded twice, with no benefit.

File: examples/segmentation_3d/assets/algo/algo.py (hunk 2 above)
Comment (jmorel) [reply examples/segmentation_3d/assets/algo/algo.py]: For each i-th item of X (`X_i`) I'm returning the prediction `y_pred_i` and the metadata associated with each `X_i`. The metadata contains the name of the file `X_i`comes from which will allow me to then be sure that the `y_i` and `y_pred_i` match the same `X_i`.

File: examples/segmentation_3d/assets/train_dataset/opener.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            ToTensor()
+        ])
+        segtrans = Compose([
+            AddChannel(),
+            RandSpatialCrop((96, 96, 96), random_size=False),
+            RandRotate90(prob=0.5, spatial_axes=(0, 2)),
+            ToTensor()
+        ])
+
+        ds = NiftiDataset(images, segs, transform=imtrans, seg_transform=segtrans)
```
Comment (Nic-Ma) [inline examples/segmentation_3d/assets/train_dataset/opener.py]: I would suggest using dict version transforms and `CacheDataset` to accelerate training.
Thanks.

**Review by Nic-Ma — COMMENTED** [review]
Thanks for your PR to combine Substra and MONAI.
I am not very familiar with Substra workflow, so mainly reviewed MONAI related parts.
I think it looks good to me except some suggestion.
Welcome to submit a PR to MONAI and add example to `examples/federated_learning/substra/`.

Thanks.

**Comment by camillemarini** [comment]
> Thanks for your PR to combine Substra and MONAI.
> I am not very familiar with Substra workflow, so mainly reviewed MONAI related parts.
> I think it looks good to me except some suggestion.
> Welcome to submit a PR to MONAI and add example to `examples/federated_learning/substra/`.
> 
> Thanks.

Ok great, we'll do !

**Comment by Nic-Ma** [comment]
Hi @camillemarini ,

We created a separate repo for FL examples: https://github.com/Project-MONAI/federated-learning-examples
And we already added MONAI Clara FL example in it. Could you please help submit to there?

Thanks.

**Comment by camillemarini** [comment]
> Hi @camillemarini ,
> 
> We created a separate repo for FL examples: https://github.com/Project-MONAI/federated-learning-examples
> And we already added MONAI Clara FL example in it. Could you please help submit to there?
> 
> Thanks.

@Nic-Ma Looks like I don't have access to it :-)

**Comment by Nic-Ma** [comment]
Oh, sorry we are still discussing the repo so it's not public so far.
Please just submit to the `examples/substra/` for review first when you are ready, then we can move to FL repo.

Thanks.

**Comment by Esadruhn** [comment]
In debug mode, in a local container with several worker, we get the following error:
`ERROR: Unexpected bus error encountered in worker. This might be caused by insufficient shared memory (shm)`

Can be solved with ([reference](https://github.com/pytorch/pytorch#docker-image))
--ipc=host or --shm-size options

Also need to check how to use the GPU with this
